If expressions update (#550)
exercism/julia: concepts.wip/if-expressions/about.md

# About

In Julia, like many other languages, allow for the following with `if`-expressions:

- Chained if-expressions
- Nested if-expressions

~~~~exercism/note

> `if`-expressions as defined here will encompass the following syntaxes:`if`, `if-else` and `if-elseif-else` statements.

Reviewer: ```suggestion
`if`-expressions as defined here will encompass the following syntaxes:`if`, `if-else` and `if-elseif-else` statements.
```

This alludes to other syntaxes existing, could you perhaps add a link to their concept pages? I'm guessing the conditionals page?

Author: Done.

~~~~

> `if`-expressions allows for branching in our programs, given a conditional is satisfied. 

Reviewer: ```suggestion
`if`-expressions allows for branching in our programs, given a condition is satisfied. 
```

Author: Done. Thanks.

# About

> In Julia, like many other languages, allow for the following with `if`-expressions:

Reviewer 2: ```suggestion
Julia, like many other languages, allows for the following `if`-expressions:
```

- Chained if-expressions
- Nested if-expressions

~~~~exercism/note
[`if`-expressions](https://docs.julialang.org/en/v1/manual/control-flow/#man-conditional-evaluation) as defined here will encompass the following syntaxes:`if`, `if-else` and `if-elseif-else` statements.
~~~~

`if`-expressions allows for branching in our programs, given a condition is satisfied.

## A first example

> The `function` defined below is not very realistic, but it shows the anatomy of an `if` statement. 

Reviewer 2: ```suggestion
The `function` defined below is not very realistic, but it shows the anatomy of an `if` expression. 
```

```julia
julia> function password_checker(password)
           if password == system_password
               return "Welcome back!"
            end
        end

# Assume system_password is "abc123"
julia> password_checker("abc123")
Welcome back!

julia> password_checker("xyz321")

```

> `println("Welcome back!")` is run, if the `password == system_password`, otherwise it is skipped, and nothing is printed to the screen.

Reviewer 2: ```suggestion
`"Welcome back!"` is returned if `password == system_password`, otherwise that block is skipped and `nothing` is returned.
```

Func returns a string now. Maybe leave out the bit about `nothing` being returned.

We can do the following for effective branching:

## Chained expressions example

```julia
# We can test for multiple cases
# with chained if expression.
if 0 ≤ n ≤ 0.5
    println("n is in the interval [0, 0.5], inclusive.")
elseif -0.5 ≤ n < 0
    println("n is in the interval [-0.5, 0).")
elseif n > 0.5
    println("n is greater than 0.5.")
else
    println("n is less than -0.5.")
end
```

> In the above chained expression, we can have mulitple `elseif` statements, with only one `else` statement at the end, but it's not necessary. 

Reviewer 2: ```suggestion
In a chained if/elseif/else expression, we can have zero or more `elseif` expressions, and at most one `else` expression at the end.
```

## Nested expressions example

```julia

> # We can refine our coniditonal evaluation

Reviewer 2: Probably better to have the comments in the markdown rather than as code comments?

- Nested if-expressions

~~~~exercism/note
[`if`-expressions](https://docs.julialang.org/en/v1/manual/control-flow/#man-conditional-evaluation) as defined here will encompass the following syntaxes:`if`, `if-else` and `if-elseif-else` statements.
~~~~

`if`-expressions allows for branching in our programs, given a condition is satisfied.

## A prompt program example

Suppose you've defined a `password_checker(password)` function that returns `true`, if the password entered by the user is correct.

```julia
# Login prompt
println("Enter Password: ")

# Get user input
password = readline()

# Our main logic
if password_checker(password)
    println("Welcome!")
end
```
Although the above program is not very realistic, but it shows the anatomy of an `if` statement. 

> In this case the `println("Welcome!")` is run, if the `password_checker(password)` returns `true`, otherwise it is skipped.

Reviewer: This is mostly a personal view, not an Exercism view, even though some other maintainers agree:

Imho any examples with reading input or printing anything should be avoided completely. So many people new to programming struggle with understanding the difference between printing and returning a value, often because their introductory classes or texts use printing everywhere. This is even more confusing in REPL settings where the difference is visually indistinguishable. To avoid this, I'd only print in an exercise that's explicitly about IO, not general examples or other exercises.

The example itself works but perhaps it could be turned into a function `password_check(password)`?

Author: Thank you for explaining this. Understood, I've converted the example into a `function`, as suggested. Please let me know if the example is clear.

Reviewer: That's clearer, though I'd change it to return a string
